src/query/deps.go

```diff
@@ -3,88 +3,75 @@ package query
 	if formatdot {
 		fmt.Fprintf(out, "digraph deps {\n")
 		fmt.Fprintf(out, "  fontname=\"Helvetica,Arial,sans-serif\"\n")
 		fmt.Fprintf(out, "  node [fontname=\"Helvetica,Arial,sans-serif\"]\n")
 		fmt.Fprintf(out, "  edge [fontname=\"Helvetica,Arial,sans-serif\"]\n")
 		fmt.Fprintf(out, "  rankdir=\"LR\"\n")
 	}
-	done := map[core.BuildLabel]bool{}
+	done := map[*core.BuildTarget]bool{}
 	for _, label := range labels {
-		if formatdot {
-			fmt.Fprintf(out, "  subgraph \"%s\" {\n", label)
-			printTargetDot(out, state, state.Graph.TargetOrDie(label), nil, done, hidden, 0, targetLevel)
-			fmt.Fprintf(out, "  }\n")
-		} else {
-			printTarget(out, state, state.Graph.TargetOrDie(label), "", done, hidden, 0, targetLevel)
-		}
+		deps(out, state, state.Graph.TargetOrDie(label), done, targetLevel, 0, hidden, formatdot)
 	}
 	if formatdot {
 		fmt.Fprintf(out, "}\n")
 	}
 }
 
-func printTarget(out io.Writer, state *core.BuildState, target *core.BuildTarget, indent string, done map[core.BuildLabel]bool, hidden bool, currentLevel int, targetLevel int) {
-	levelLimitReached := targetLevel != -1 && currentLevel == targetLevel
-	if done[target.Label] || levelLimitReached {
+// deps looks at all the deps of the given target & recurses into them, printing as appropriate.
+func deps(out io.Writer, state *core.BuildState, target *core.BuildTarget, done map[*core.BuildTarget]bool, targetLevel, currentLevel int, hidden, formatdot bool) {
+	if currentLevel == targetLevel {
 		return
 	}
-
-	if state.ShouldInclude(target) && (hidden || !target.HasParent()) {
-		fmt.Fprintf(out, "%s%s\n", indent, target)
-		indent += "  "
-		currentLevel++
-	}
-	done[target.Label] = true
-
-	for _, dep := range target.Dependencies() {
-		printTarget(out, state, dep, indent, done, hidden, currentLevel, targetLevel)
-	}
-	if target.Subrepo != nil && target.Subrepo.Target != nil {
-		printTarget(out, state, target.Subrepo.Target, indent, done, hidden, currentLevel, targetLevel)
-	}
-}
-
-func printTargetDot(out io.Writer, state *core.BuildState, target *core.BuildTarget, parent *core.BuildTarget, done map[core.BuildLabel]bool, hidden bool, currentLevel int, targetLevel int) {
-	levelLimitReached := targetLevel != -1 && currentLevel == targetLevel
-
-	if state.ShouldInclude(target) && (hidden || !target.HasParent()) {
-		if !done[target.Label] {
-			shape := "ellipse"
-			if target.IsFilegroup {
-				shape = "folder"
-			} else if target.IsRemoteFile {
-				shape = "octagon"
-			} else if target.IsTextFile {
-				shape = "note"
-			} else if target.IsBinary {
-				shape = "component"
-			}
-			fmt.Fprintf(out, "   node [shape=%s] \"%s\";\n", shape, target)
+	for _, l := range target.DeclaredDependencies() {
+		dep := state.Graph.TargetOrDie(l)
+		if !state.ShouldInclude(dep) || done[dep] {
+			continue // target is filtered out
 		}
-		if parent != nil {
-			fmt.Fprintf(out, "   \"%s\" -> \"%s\";\n", parent, target)
+		done[dep] = true
+		if hidden || !dep.HasParent() {
```

Comment:
I can't understand why whether the dep has no parent is the condition for whether it should be printed... I know this hasn't changed but a comment might help.